### PR TITLE
[libc++abi] Use __has_feature check to enable usage of thread_local for exception storage

### LINKED ISSUE #78207: [libcxxabi] HAS_THREAD_LOCAL is not defined by cmake
`libcxxabi/src/cxa_exception_storage.cpp` contain a check for `HAS_THREAD_LOCAL`, but the flags is set by `compiler-rt/lib/fuzzer/CMakeLists.txt`. This cmake script is not used in normal build of libcxxabi.

Maybe it should be replaced by `__has_feature(cxx_thread_local)`?

See also: https://reviews.llvm.org/D155278

### PR DESCRIPTION
**Pull request by ldionne**
Previously, we'd use HAS_THREAD_LOCAL which was never defined. Hence, we'd basically never use the code path where we use thread_local.

Fixes #78207

### CODE REVIEW

**Review by MaskRay — APPROVED** [review]
LGTM. @mstorsjo may want to take a look https://github.com/llvm/llvm-project/pull/79542#issuecomment-1913361802

**Comment by mstorsjo** [comment]
> LGTM. @mstorsjo may want to take a look [#79542 (comment)](https://github.com/llvm/llvm-project/pull/79542#issuecomment-1913361802)

Thanks; I tested this patch in some of my test environments, and it seems to not break anything, so I think it's fine.

As mentioned in https://github.com/llvm/llvm-project/pull/79542#issuecomment-1913361802, this could have an effect on emutls environments; I tried to run a set of tests with emutls enabled as well. With emutls enabled (as in that linked PR), a bunch of my tests hang, both with and without this PR. So that doesn't indicate any reason to hold back this patch either.

So, in short, +1 for this patch from me, and AFAIK this will fix the same issue as https://reviews.llvm.org/D155278 tried to fix.

**Comment by nico** [comment]
In case anyone else wonders why their binaries that statically link libc++abi no longer `dlclose()` after this change, https://crbug.com/351867703 has some details on that.

(No action needed for libc++abi, just mentioning it in case someone else runs into this.)
